File: oop/python/lista01/Ex08.py

from KeyboardInput import KeyboardInput


class Ex08:
    def __init__(self) -> None:
        self.ERROR = 10e-7
        self.__iterator = 0

    def calculateBaseEquation(self, x) -> float: return x**3 - x**2 - 13*x + 8

    def getLastOperationTotalIterations(self) -> int: return self.__iterator

    def solveByBisection(self, a, b) -> float:
        if self.calculateBaseEquation(a) * self.calculateBaseEquation(b) >= 0: return a

        c = a
        self.__iterator = 0
        while b - a >= self.ERROR:
            c = (a + b) / 2
            if (self.calculateBaseEquation(c) == 0.0): break
            elif self.calculateBaseEquation(c) * self.calculateBaseEquation(a) < 0: b = c
            else: a = c

            self.__iterator += 1
        return c

if __name__ == "__main__":
    bisectionCalculator = Ex08()
    keyboardInput = KeyboardInput()

    print("Provide the desired interval: ")
    a = keyboardInput.readFloat()
    b = keyboardInput.readFloat()

    print("Solution by Bisection: {}".format(bisectionCalculator.solveByBisection(a, b)))
    print("{} total iterations".format(bisectionCalculator.getLastOperationTotalIterations()))

File: oop/lista01_java/Ex08/Ex08.java
package Lista01.Ex08;

import Lista01.KeyboardInput;

public class Ex08 {
    static final Double ERROR = 10e-7;
    private Integer iterator = 0;

    private Double calculateBaseEquation(Double x) {
        return Math.pow(x, 3) - Math.pow(x, 2) - 13*x + 8;
    }

    public Integer getLastOperationTotalIterations() { return this.iterator; }

    public Double solveByBisection(Double a, Double b) {
        if (calculateBaseEquation(a) * calculateBaseEquation(b) >= 0) return a;

        Double c = a;
        this.iterator = 0;
        while (b - a >= ERROR) {
            c = (a + b) / 2;
            if (calculateBaseEquation(c) == 0.0) break;
            else if (calculateBaseEquation(c) * calculateBaseEquation(a) < 0) b = c;
            else a = c;

            this.iterator++;
        }

        return c;
    }

    public static void main(String[] args) {
        Ex08 bisectionCalculator = new Ex08();
        KeyboardInput keyboardInput = new KeyboardInput(System.in);

        System.out.println("Provide the desired interval: ");
        Double a = keyboardInput.readDouble();
        Double b = keyboardInput.readDouble();

        System.out.printf("Solution by Bisection: %.4f\n", bisectionCalculator.solveByBisection(a, b));
        System.out.printf("%d total iterations", bisectionCalculator.getLastOperationTotalIterations());
    }
}
